crates/core/src/host/host_controller.rs

```diff
@@ -46,37 +48,57 @@ type HostCell = Arc<AsyncRwLock<Option<Host>>>;
 /// The registry of all running hosts.
 type Hosts = Arc<Mutex<IntMap<u64, HostCell>>>;
 
-type ExternalStorage = dyn Fn(&Hash) -> anyhow::Result<Option<AnyBytes>> + Send + Sync + 'static;
-type SameDbStorage = dyn Fn(&RelationalDB, &Hash) -> anyhow::Result<Option<AnyBytes>> + Send + Sync + 'static;
+#[async_trait]
+pub trait ExternalStorage: Send + Sync + 'static {
+    async fn lookup(&self, program_hash: Hash) -> anyhow::Result<Option<AnyBytes>>;
+}
+#[async_trait]
+impl<F, Fut> ExternalStorage for F
+where
+    F: Fn(Hash) -> Fut + Send + Sync + 'static,
+    Fut: Future<Output = anyhow::Result<Option<AnyBytes>>> + Send,
+{
+    async fn lookup(&self, program_hash: Hash) -> anyhow::Result<Option<AnyBytes>> {
+        self(program_hash).await
+    }
+}
+
+type SameDbStorage = dyn Fn(&RelationalDB, Hash) -> anyhow::Result<Option<AnyBytes>> + Send + Sync + 'static;
 
 /// Source of the module binary.
 ///
 /// This is a temporary measure until modules are stored in their database itself.
 #[derive(Clone)]
 pub enum ProgramStorage {
     /// External storage, addressable by program hash.
-    External(Arc<ExternalStorage>),
+    External(Arc<dyn ExternalStorage>),
     /// Storage inside the [`RelationalDB`] itself, but with a retrieval
     /// function chosen by the user.
     SameDb(Arc<SameDbStorage>),
 }
 
 impl ProgramStorage {
     /// Construct a [`ProgramStorage::External`] from a lookup function.
-    pub fn external<F>(f: F) -> Self
-    where
-        F: Fn(&Hash) -> anyhow::Result<Option<AnyBytes>> + Send + Sync + 'static,
-    {
-        Self::External(Arc::new(f))
+    pub fn external(s: impl ExternalStorage) -> Self {
+        Self::External(Arc::new(s))
     }
 
     /// Construct a [`ProgramStorage::SameDb`] from a lookup function.
     pub fn same_db<F>(f: F) -> Self
     where
-        F: Fn(&RelationalDB, &Hash) -> anyhow::Result<Option<AnyBytes>> + Send + Sync + 'static,
+        F: Fn(&RelationalDB, Hash) -> anyhow::Result<Option<AnyBytes>> + Send + Sync + 'static,
     {
         Self::SameDb(Arc::new(f))
     }
+
+    async fn load_program(&self, db: &RelationalDB, program_hash: Hash) -> anyhow::Result<AnyBytes> {
+        debug!("lookup program {}", program_hash);
+        match self {
+            ProgramStorage::External(external) => external.lookup(program_hash).await,
+            ProgramStorage::SameDb(f) => f(db, program_hash),
```

Comment:
Wondering if this should be spawn blocking, just in case the same thread has a tx open…

Reply:
Hmm, probably - but we can cross that when we get to actually using it.